File: backend/app/protocol.py
"""Wire protocol: one JSON envelope both ways, a frozen message catalog.

Server -> client: {"t": type, "seq": int, "match": match_id|None, "ts": epoch_ms, "d": payload}
Client -> server: {"t": type, "cseq": int, "d": payload}

Motion sample layout (13 numbers): [t_phone_ms, ax, ay, az, agx, agy, agz, rx, ry, rz, alpha, beta, gamma]
  a*  = acceleration without gravity (m/s^2), ag* = including gravity, r* = rotationRate (deg/s),
  alpha/beta/gamma = orientation angles (deg). t_phone_ms = performance.timeOrigin + event.timeStamp.
"""
from __future__ import annotations

import json
from typing import Any, Literal

from pydantic import BaseModel, ConfigDict, Field, ValidationError

SAMPLE_LEN = 13
Role = Literal["remote", "rail", "projector", "host"]
ROLES: tuple[str, ...] = ("remote", "rail", "projector", "host")


class Payload(BaseModel):
    model_config = ConfigDict(extra="allow")


class Hello(Payload):
    role: Role
    device_id: str = Field(min_length=4, max_length=64)
    token: str | None = None
    nickname: str | None = Field(default=None, max_length=24)
    ua: str | None = Field(default=None, max_length=300)
    time_origin: float | None = None
    last_seq: int | None = None


class Ping(Payload):
    t_client: float
    rtt_ms: float = 0.0


class Resync(Payload):
    last_seq: int = 0


class MotionFrame(Payload):
    t0: float
    n: int
    s: list[list[float]]


class Bet(Payload):
    market_id: str
    outcome_id: str
    stake: int = Field(ge=1)


class SponsorBuy(Payload):
    move_id: str
    target: Literal["house", "human"] | None = None


class CardVote(Payload):
    pairing_id: str


class CrateBid(Payload):
    amount: int = Field(ge=1)


class PairRequest(Payload):
    kind: Literal["pass_seat", "transfer", "bind_glove", "tag"]
    amount: int | None = None
    seat_id: str | None = None


class SetNickname(Payload):
    nickname: str = Field(min_length=1, max_length=24)


class InputAction(Payload):
    """Direct input from a keyboard remote (desktop testing): bypasses the motion detectors."""
    kind: Literal["swing", "release", "punch", "block_on", "block_off", "dodge", "parry", "move", "bump", "shake", "flick"]
    params: dict[str, Any] = Field(default_factory=dict)
    t_client: float | None = None


# Every client message type and the model that validates its payload.
CLIENT_MESSAGES: dict[str, type[Payload]] = {
    "session.hello": Hello,
    "session.ping": Ping,
    "session.resync": Resync,
    "session.nickname": SetNickname,
    "motion.frame": MotionFrame,
    "motion.calib_done": Payload,
    "input.action": InputAction,
    "market.bet": Bet,
    "sponsor.buy": SponsorBuy,
    "crate.bid": CrateBid,
    "card.vote": CardVote,
    "pair.request": PairRequest,
    "host.telemetry": Payload,
    "host.start": Payload,
    "host.pause": Payload,
    "host.resume": Payload,
    "host.next": Payload,
    "host.force_scenario": Payload,
    "host.set_param": Payload,
    "host.kick": Payload,
    "host.release_seat": Payload,
    "host.lock_seat": Payload,
    "host.set_public_url": Payload,
    "host.reload_config": Payload,
    "host.toggle": Payload,
    "host.set_seats": Payload,
    "host.card": Payload,
    "host.unlock_audio": Payload,
    "host.adjust_chips": Payload,
    "host.set_backend_url": Payload,
}

# Every server message type. Kept as a flat tuple so scripts/check_protocol.py can diff it against protocol.ts.
SERVER_MESSAGES: tuple[str, ...] = (
    "session.welcome", "session.pong", "session.snapshot", "session.error",
    "motion.status", "motion.meter", "motion.gesture", "motion.calib",
    "seat.update",
    "match.start", "match.phase", "match.tick", "match.turn_result", "match.end", "match.pause",
    "agent.decision", "agent.studying",
    "market.window", "market.bet_ack", "market.odds", "market.settle", "market.leaderboard", "market.balance",
    "sponsor.warn", "sponsor.applied", "sponsor.ack", "crate.open", "crate.result", "card.vote_open", "card.vote_result",
    "pair.prompt", "pair.result",
    "voice.line", "sfx.play",
    "host.ack", "host.config", "host.diagnostics", "ladder.update",
)

# Types the per-connection send queue may drop first when a client is slow (high-rate, superseded by the next one).
DROPPABLE: frozenset[str] = frozenset({"motion.meter", "market.odds", "match.tick", "motion.status"})


class ClientMsg(BaseModel):
    t: str
    cseq: int = 0
    d: dict[str, Any] = Field(default_factory=dict)


class Envelope(BaseModel):
    t: str
    seq: int
    match: str | None = None
    ts: int
    d: dict[str, Any] = Field(default_factory=dict)

    def dumps(self) -> str:
        return json.dumps(self.model_dump(), separators=(",", ":"), default=_json_default)


def _json_default(o):
    if hasattr(o, "model_dump"):
        return o.model_dump()
    if hasattr(o, "__dict__"):
        return o.__dict__
    raise TypeError(f"not serializable: {type(o)}")


class ProtocolError(ValueError):
    pass


def parse_client(raw: str | bytes) -> ClientMsg:
    """Validate an inbound message: known type, payload shape. Raises ProtocolError."""
    try:
        obj = json.loads(raw)
    except (ValueError, TypeError) as e:
        raise ProtocolError(f"bad json: {e}") from e
    if not isinstance(obj, dict) or "t" not in obj:
        raise ProtocolError("missing t")
    t = obj["t"]
    model = CLIENT_MESSAGES.get(t)
    if model is None:
        raise ProtocolError(f"unknown type {t!r}")
    d = obj.get("d") or {}
    try:
        payload = model.model_validate(d)
    except ValidationError as e:
        raise ProtocolError(f"{t}: {e.errors()[0]['msg']} at {e.errors()[0]['loc']}") from e
    if t == "motion.frame":
        for s in payload.s:  # type: ignore[attr-defined]
            if len(s) != SAMPLE_LEN:
                raise ProtocolError(f"motion.frame sample must have {SAMPLE_LEN} numbers")
    return ClientMsg(t=t, cseq=int(obj.get("cseq", 0) or 0), d=payload.model_dump())

File: web/src/protocol.ts
// Hand-written mirror of backend/app/protocol.py. backend/scripts/check_protocol.py diffs the type lists.
export const SERVER_MESSAGES = [
  'session.welcome', 'session.pong', 'session.snapshot', 'session.error',
  'motion.status', 'motion.meter', 'motion.gesture', 'motion.calib',
  'seat.update',
  'match.start', 'match.phase', 'match.tick', 'match.turn_result', 'match.end', 'match.pause',
  'agent.decision', 'agent.studying',
  'market.window', 'market.bet_ack', 'market.odds', 'market.settle', 'market.leaderboard', 'market.balance',
  'sponsor.warn', 'sponsor.applied', 'sponsor.ack', 'crate.open', 'crate.result', 'card.vote_open', 'card.vote_result',
  'pair.prompt', 'pair.result',
  'voice.line', 'sfx.play',
  'host.ack', 'host.config', 'host.diagnostics', 'ladder.update',
] as const

export const CLIENT_MESSAGES = [
  'session.hello', 'session.ping', 'session.resync', 'session.nickname',
  'motion.frame', 'motion.calib_done', 'input.action',
  'market.bet', 'sponsor.buy', 'crate.bid', 'card.vote', 'pair.request',
  'host.telemetry', 'host.start', 'host.pause', 'host.resume', 'host.next', 'host.force_scenario', 'host.set_param',
  'host.kick', 'host.release_seat', 'host.lock_seat', 'host.set_public_url', 'host.reload_config', 'host.toggle',
  'host.set_seats', 'host.card', 'host.unlock_audio', 'host.adjust_chips',
] as const

export type ServerType = (typeof SERVER_MESSAGES)[number]
export type ClientType = (typeof CLIENT_MESSAGES)[number]
export type Role = 'remote' | 'rail' | 'projector' | 'host'

// Motion sample: [t_phone_ms, ax, ay, az, agx, agy, agz, rx, ry, rz, alpha, beta, gamma]
export type Sample = [number, number, number, number, number, number, number, number, number, number, number, number, number]
export const SAMPLE_LEN = 13

export interface Envelope<T = Record<string, unknown>> { t: ServerType; seq: number; match: string | null; ts: number; d: T }

export interface Seat { seat_id: string; label: string; status: 'open' | 'claimed' | 'locked'; device_id: string | null; nickname: string | null; hand: string | null; side: string | null; join_url: string | null }
export interface Welcome { device_id: string; role: Role; nickname: string | null; seat_id: string | null; seat_token: string | null; seat_label: string | null; hand: string | null; reason: string | null; public_url: string; rail_url: string; server_ts: number; mode: string; dev: boolean; balance: number }
export interface DeviceInfo { device_id: string; role: Role; nickname: string | null; connected: boolean; seat_id: string | null; hz: number; offset_ms: number; calibrated: boolean; last_gesture: string | null }
export interface Outcome { id: string; label: string; pool: number }
export interface Market { market_id: string; kind: string; label: string; outcomes: Outcome[]; closes_ts: number; open: boolean; turn_no?: number; winner?: string | string[] | null; status?: string; pool?: number; match_id?: string | null }
export interface MatchSummary { match_id: string; sport: 'bowling' | 'baseball' | 'boxing'; seed: number; human_seats: string[]; opponent: { tier: string; name: string; accent?: string }; players?: Record<string, string>; phase?: string; turn_no?: number; deadline_ts?: number; score?: Record<string, unknown>; paused?: boolean; card?: boolean; fighters?: Record<string, unknown> }
export interface LeaderRow { device_id: string; nickname: string; chips: number; titles: string[] }
export interface Snapshot { public_url: string; rail_url: string; seats: Seat[]; match: MatchSummary | null; markets: Market[]; leaderboard: LeaderRow[]; ladder: Record<string, Record<string, number>>; studying: Record<string, unknown>; card: Record<string, unknown> | null; crate: Record<string, unknown> | null; toggles: Record<string, boolean>; me?: { device_id: string; balance: number; nickname?: string | null; seat_id?: string | null; calibrated?: boolean }; devices?: DeviceInfo[]; seat_tokens?: Record<string, string>; audio_unlocked?: boolean }
export interface Gesture { seat_id: string | null; device_id: string; kind: string; t_phone: number; t_server: number; power: number; axis: string; sign: number; duration_ms: number; extra?: Record<string, unknown> }
export interface VoiceLine { priority: number; speaker: string; text: string; url: string | null; duration_ms: number; id: string }
